lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -542,17 +542,22 @@ def update(self, trans: GalaxyWebTransaction, id, payload, **kwds):
                 trans.sa_session.flush()
 
             if "menu_entry" in workflow_dict or "show_in_tool_panel" in workflow_dict:
-                if workflow_dict.get("menu_entry") or workflow_dict.get("show_in_tool_panel"):
-                    workflow_ids = [wf.stored_workflow_id for wf in trans.user.stored_workflow_menu_entries]
-                    if trans.security.decode_id(id) not in workflow_ids:
+                show_in_panel = workflow_dict.get("menu_entry") or workflow_dict.get("show_in_tool_panel")
+                stored_workflow_menu_entries = trans.user.stored_workflow_menu_entries
+                decoded_id = trans.security.decode_id(id)
+                if show_in_panel:
+                    workflow_ids = [wf.stored_workflow_id for wf in stored_workflow_menu_entries]
+                    if decoded_id not in workflow_ids:
                         menuEntry = model.StoredWorkflowMenuEntry()
                         menuEntry.stored_workflow = stored_workflow
-                        trans.user.stored_workflow_menu_entries.append(menuEntry)
+                        stored_workflow_menu_entries.append(menuEntry)
                 else:
                     # remove if in list
-                    entries = {x.stored_workflow_id: x for x in trans.user.stored_workflow_menu_entries}
-                    if trans.security.decode_id(id) in entries:
-                        trans.user.stored_workflow_menu_entries.remove(entries[trans.security.decode_id(id)])
+                    entries = {x.stored_workflow_id: x for x in stored_workflow_menu_entries}
+                    if decoded_id in entries:
+                        stored_workflow_menu_entries.remove(entries[decoded_id])
+                # flush is required is if no other attributes are being updated...
+                trans.sa_session.flush()
```

Comment:
The comment could be read as not needing a flush if annotation/importable/etc are being updated. Maybe drop the comment, it should be clear why we need a flush here.

```suggestion
                trans.sa_session.flush()
```
Maybe also explicitly add `menuEntry` to the session ?

Any reason not to flush once at the end of the method ? That's probably better than the individual flushes above an here.

Reply:
Hmm... I've deleted so many unneeded flushes I think I tend assume all the flushes are unneeded 😆  so I thought a required flush was worthy of a comment. I will rework this a bit based on your comments though.